src/codelabs/async-await/index.md

```diff
@@ -0,0 +1,346 @@
+---
+title: Asynchronous Dart code - futures, async, and await 
+description: A series of explanations and exercises for writing asynchronous code in Dart
+date: 2019-06-12
+tags: [ future, async, await, asynchronous, try, catch, error ]
+---
+
+## Introduction
+ 
+Welcome to the asynchronous Dart codelab! In this codelab you practice using
+Dart to run asynchronous code via the `async` and `await` keywords.
+
+In addition to explanations, this codelab includes embedded editors that show 
+you working code examples and coding exercises. You can use these editors to 
+test your knowledge by completing the exercises.
+
+Before you begin this codelab, you should:
+
+* Have programming experience in another language
+* Know basic syntax in Dart 
+* Have some experience writing asynchronous code in another language
+
+When you finish this codelab, you'll know the following: 
+
+* How to use the `async` keyword
+* How to use the `await` keyword
+* When to use `async` and `await` in your code
+* How your code executes when using the `async/await` keywords
+* How to handle errors from an asynchronous call using `try/catch` expressions 
+in async functions.
+* Where to find more information for the `async/await` material 
+covered in this codelab.
+
+
+## Why async matters
+
+Asynchronous operations let your program complete other work while waiting for 
+an operation to finish. Some common processes that require asynchronous 
+operations include: 
+
+* Fetching data over a network
+* Writing to a database
+* Reading data from a file
+
+To implement these features, you will need to learn the Dart syntax for
+futures, async, and await. In Dart you cannot use synchronous code to handle
+asynchronous operations.
+
+Consider the following example that fails to print the desired user order to the
+console: 
+
+
+<iframe style="margin-top: 25px; margin-bottom: 25px" frameborder="no" height="420" src="https://dartpad.dartlang.org/experimental/embed-new.html?id=5c8c7716b6b4284842f15fe079f61e47" width="100%"></iframe>
+
+
+In the above code:
+
+* `getUserOrder` is an asynchronous function that waits four seconds before
+returning a string describing the user's order: a large latte. (For a further
+explanation you can [find out more about
+Future.delayed]({{site.dart_api}}/Future.delayed.html))
+* `createOrderMessage` calls `getUserOrder` as if it were a synchronous function
+ -- it invokes `getUserOrder` and immediately prints the result. 
+* Since `getUserOrder` waits four seconds before returning the user's order,
+`createOrderMessage` never actually prints the user's order. Instead the console
+prints the value immediately returned by `createOrderMessage` which
+turns out to be `Instance of '_Future<String>'`.
+
+What is a future, and how do you write code to handle the asynchronous function
```

Comment:
What is a future, and how do you write code to handle the asynchronous function
 `getUserOrder`?

**wait, I'm lost. you ask what is a future above, and then below (in quick review) you go over how Futures write declarative asynchronous code. is a future different from Futures? are they terms uses synonymously? did you go over Futures before quick review?**

 {{ site.alert.secondary }} 
 Quick review: 
 * Asynchronous operations let a program do work while waiting for an operation to finish.
 * In Dart, you can use Futures, `async` and `await` to write declarative asynchronous code.
 {{ site.alert.end }}

Reply:
I'll remove mention of futures from the quick review. They're covered in the next section.